dropshot/src/server.rs

```diff
@@ -149,49 +225,294 @@ impl<C: ServerContext> HttpServerStarter<C> {
+        pinned.poll_flush(ctx)
+    }
+
+    fn poll_shutdown(
+        mut self: Pin<&mut Self>,
+        ctx: &mut core::task::Context,
+    ) -> Poll<std::io::Result<()>> {
+        let pinned = Pin::new(&mut self.stream);
+        pinned.poll_shutdown(ctx)
+    }
+}
+
+/// This is our bridge between tokio-rustls and hyper. It implements
+/// `hyper::server::accept::Accept` interface, producing TLS-over-TCP
+/// connections.
+///
+/// Internally, it creates a stream that produces fully negotiated TLS
+/// connections as they come in from a TCP listen socket.  This stream allows
+/// for multiple TLS connections to be negotiated concurrently with new
+/// connections being accepted.
+struct HttpsAcceptor {
+    stream: Box<dyn Stream<Item = std::io::Result<TlsConn>> + Send + Unpin>,
+}
+
+impl HttpsAcceptor {
+    pub fn new(
+        log: slog::Logger,
+        tls_acceptor: TlsAcceptor,
+        tcp_listener: TcpListener,
+    ) -> HttpsAcceptor {
+        HttpsAcceptor {
+            stream: Box::new(Box::pin(Self::new_stream(
+                log,
+                tls_acceptor,
+                tcp_listener,
+            ))),
+        }
+    }
+
+    fn new_stream(
+        log: slog::Logger,
+        tls_acceptor: TlsAcceptor,
+        tcp_listener: TcpListener,
+    ) -> impl Stream<Item = std::io::Result<TlsConn>> {
+        stream! {
+            let mut tls_negotiations = futures::stream::FuturesUnordered::new();
+            loop {
+                tokio::select! {
+                    Some(negotiation) = tls_negotiations.next(), if
+                            !tls_negotiations.is_empty() => {
+
+                        match negotiation {
+                            Ok(conn) => yield Ok(conn),
+                            Err(e) => {
+                                // If TLS negotiation fails, log the cause but
+                                // don't forward it along. Yielding an error
+                                // from here will terminate the server.
+                                // TODO: We may want to export a counter for
+                                // different error types, since this may contain
+                                // useful things like "your certificate is
+                                // invalid"
+                                info!(log, "tls accept err: {}", e);
+                            },
+                        }
+                    },
+                    accept_result = tcp_listener.accept() => {
+                        let (socket, addr) = match accept_result {
+                            Ok(v) => v,
+                            Err(e) => {
+                                match e.kind() {
+                                    std::io::ErrorKind::ConnectionAborted => {
+                                        continue;
+                                    },
+                                    // The other errors that can be returned
```

Comment:
Can we add a TODO-resilience or TODO-robustness or something here?  I want to make sure we look at this when it comes to hardening.